system/mp/test/mptest.cpp

```diff
@@ -701,6 +704,411 @@ void MPAlltoAll(IGroup *group, ICommunicator *mpicomm, size32_t buffsize=0, unsi
+        }
+        const char *queryTgtRanks() { return tgtRanksStr; }
+        virtual void threadmain() override
+        {
+            passed = false;
+            try
+            {
+                CMessageBuffer msg, recvMsg;
+                msg.setReplyTag(replyTag);
+                void *data = msg.reserveTruncate(msgSize);
+                unsigned hash = (unsigned)mpTag;
+
+                VStringBuffer logMsg("NxN: mpTag=%u, dstRank(s) [%s]", (unsigned)mpTag, tgtRanksStr.str());
+                PROGLOG("%s", logMsg.str());
+
+                unsigned __int64 remaining = totalSendSize;
+                CCycleTimer timer;
+                while (true)
+                {
+                    size32_t sz;
+                    if (remaining >= msgSize)
+                    {
+                        sz = msgSize;
+                        remaining -= msgSize;
+                    }
+                    else
+                    {
+                        sz = remaining;
+                        msg.setLength(sz);
+                        remaining = 0;
+                    }
+                    hash = fillData(data, sz, hash);
+                    for (rank_t t: tgtRanks)
+                    {
+                        if (async)
+                            ackThread.addHash(t, hash);
+                        if (!comm->send(msg, t, mpTag))
+                            throwUnexpected();
+                    }
+                    if (!async)
+                    {
+                        rank_t sender;
+                        for (int t: tgtRanks)
+                        {
+                            rank_t sender;
+                            unsigned receivedHash;
+                            assertex(receiveAck(recvMsg, sender, receivedHash));
+                            verifyAck(hash, receivedHash);
+                        }
+                    }
+                    if (!remaining)
+                        break;
+                }
+                msg.clear();
+                // send blank msg to all to signal end to receivers.
+                if (!comm->send(msg, RANK_ALL, mpTag))
+                    throwUnexpected();
+                if (async)
+                    ackThread.join();
+                else
+                {
+                    rank_t sender;
+                    for (unsigned r=0; r<grpSize; r++)
+                    {
+                        while (!comm->recv(msg, r, replyTag, &sender, 60000))
+                            WARNLOG("Waiting for final ack");
+                        assertex(sender == r);
+                        assertex(0 == msg.length());
+                    }
+                }
+                float ms = timer.elapsedMs();
+                float mbPerSec = (totalSendSize/ms*1000)/0x100000;
+                PROGLOG("Stream stats: time taken = %.2f seconds, total sent=%u MB, throughput = %.2f MB/s", ms/1000, (unsigned)(totalSendSize/0x100000), mbPerSec);
```

Comment:
I'm never sure, is the ms/1000 done with floating pt by default ? Or should it be (float)ms/1000.0 ?

Reply:
ok, ms is float so I think yes

Reply:
as long as one side is float, result will be float.